src/buildHTML.js

```diff
@@ -17,18 +17,56 @@ var utils = require("./utils");
 
 var makeSpan = buildCommon.makeSpan;
 
+var isSpace = function(node) {
+    return node instanceof domTree.span && node.classes[0] === "mspace";
+};
+
 /**
  * Take a list of nodes, build them in order, and return a list of the built
  * nodes. This function handles the `prev` node correctly, and passes the
- * previous element from the list as the prev of the next element.
+ * previous element from the list as the prev of the next element, ignoring
+ * spaces. documentFragments are flattened into their contents, so the
+ * returned list contains no fragments.
  */
 var buildExpression = function(expression, options, prev) {
+    // Parse expressions into `groups`.
     var groups = [];
     for (var i = 0; i < expression.length; i++) {
         var group = expression[i];
-        groups.push(buildGroup(group, options, prev));
-        prev = group;
+        var output = buildGroup(group, options, prev);
+        if (output instanceof domTree.documentFragment) {
+            Array.prototype.push.apply(groups, output.children);
+        } else {
+            groups.push(output);
+        }
+        if (!isSpace(output)) {
+            prev = group;
+        }
     }
+    // At this point `groups` consists entirely of `symbolNode`s and `span`s.
+
+    // Explicit spaces (e.g., \;, \,) should be ignored with respect to atom
+    // spacing (e.g., "add thick space between mord and mrel"). Since CSS
+    // adjacency rules implement atom spacing, spaces should be invisible to
+    // CSS. So we shift them into the atoms themselves.
+    var lastSpace = -1;
+    for (i = 0; i < groups.length; i++) {
+        if (isSpace(groups[i])) {
+            if (lastSpace === -1) {
+                lastSpace = i;
+            }
+        } else if (lastSpace >= 0) {
+            var node = groups[i];
+            if (node instanceof domTree.symbolNode) {
+                node = groups[i] = makeSpan(node.classes, [node]);
+            }
+            buildCommon.prependChildren(node,
+                groups.splice(lastSpace, i - lastSpace));
```

Comment:
I found this kind of hard to follow.  I think maybe doing two passes would make this more understandable, in the first pass we could add divide `groups` up into set of arrays and the second pass could create the spans.  This should avoid the need for `prependChildren`.